dpctl/dparray.py

```diff
@@ -0,0 +1,257 @@
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buf,
+                offset=0,
+                strides=strides,
+                order=order,
+            )
+            np.copyto(new_obj, ar, casting="no")
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+
+    def __array_finalize__(self, obj):
+        dprint("__array_finalize__:", obj, hex(id(obj)), type(obj))
+        # When called from the explicit constructor, obj is None
+        if obj is None:
+            return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
+            mobj = obj
+            while isinstance(mobj, numba.core.runtime._nrt_python._MemInfo):
+                dprint("array_finalize got Numba MemInfo")
+                ea = mobj.external_allocator
+                d = mobj.data
+                dprint("external_allocator:", hex(ea), type(ea))
+                dprint("data:", hex(d), type(d))
+                dppl_rt_allocator = numba.dppl._dppl_rt.get_external_allocator()
+                dprint(
+                    "dppl external_allocator:",
+                    hex(dppl_rt_allocator),
+                    type(dppl_rt_allocator),
+                )
+                dprint(dir(mobj))
+                if ea == dppl_rt_allocator:
+                    return
+                mobj = mobj.parent
+                if isinstance(mobj, ndarray):
+                    mobj = mobj.base
+        if isinstance(obj, np.ndarray):
+            ob = self
+            while isinstance(ob, np.ndarray):
+                if hasattr(obj, array_interface_property):
+                    return
+                ob = ob.base
+
+        # Just raise an exception since __array_ufunc__ makes all reasonable cases not
+        # need the code below.
+        raise ValueError(
+            "Non-USM allocated ndarray can not viewed as a USM-allocated one without a copy"
+        )
+
+    # Tell Numba to not treat this type just like a NumPy ndarray but to propagate its type.
+    # This way it will use the custom dparray allocator.
+    __numba_no_subtype_ndarray__ = True
+
+    # Convert to a NumPy ndarray.
+    def as_ndarray(self):
+        return np.copy(self)
+
+    def __array__(self):
+        return self
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        if method == "__call__":
```

Comment:
Need to also handle `method="accumulate"` , `"reduce"`, `"reduceat"`, `"outer"`

Reply:
Can we add a TODO and open a ticket to track this future work.